### ISSUE TITLE
WorkflowFunction Class Refactor

### ISSUE THREAD

**Issue by SoymilkWinsAgain** [TARPS-group/prob-pipe#196]
## Summary

`WorkflowFunction` is the central execution facade for `ProbPipe` operations and user-defined workflow functions. It should remain the public facade, but it should no longer be the implementation hub for call resolution, distribution normalization, broadcast planning, sampling, sweep execution, JAX vectorization, execution dispatch, output coercion, and provenance assembly. In a word, this class is turning into a big ball of mud.

This plan proposes a staged internal refactor that keeps the public API stable while moving distinct responsibilities into private core modules. The immediate goal is maintainability and testability. The longer-term goal is to expose a **backend-neutral** internal execution interface: execution adapters should receive plain functions, plain call dictionaries, and resolved execution configuration rather than a `WorkflowFunction` instance.

## Status Quo of `WorkflowFunction`

`WorkflowFunction` currently does much more than wrap and call a user function. It is the public facade for user-defined workflow functions, but it also owns call resolution, type normalization, broadcast planning, sampling, sweep execution, local vectorization, backend dispatch, output coercion, and provenance assembly. The result is a class where unrelated changes require a maintainer to reason across several implementation domains at once.

### Current Attributes

| Category | Attribute | Source | Current responsibility |
|---|---|---|---|
| Distribution broadcast | `DEFAULT_N_BROADCAST_SAMPLES` | Class attribute | Default distribution broadcast sample count, currently `128`. |
| Public facade and call resolution | `_func` | `__init__` argument `func` | Wrapped user function. |
| Public facade and call resolution | `_sig` | `inspect.signature(func)` | Function signature used for positional binding, defaults, and validation. |
| Public facade and call resolution | `_hints` | `get_type_hints(func)` | Type hints used to identify dependencies, distributions, protocols, and array behavior. |
| Execution dispatch | `_workflow_kind_raw` | `__init__` argument `workflow_kind` | Per-instance Prefect orchestration setting before global fallback resolution. |
| Public facade and call resolution | `_name` | `__init__` argument `name` or `func.__name__` | Display and operation name used by results, provenance, and orchestration labels. |
| Public facade and call resolution | `__doc__` | Copied from `func.__doc__` | Introspection metadata for docs and help output. |
| Public facade and call resolution | `__name__` | Set from `_name` | Callable name exposed to introspection. |
| Public facade and call resolution | `__qualname__` | Copied from `func.__qualname__` or `_name` | Qualified callable name exposed to introspection. |
| Public facade and call resolution | `__signature__` | Set from `_sig` | Makes `inspect.signature()` report the wrapped function signature. |
| Public facade and call resolution | `__module__` | Copied from `func.__module__` | Module metadata exposed to introspection. |
| Public facade and call resolution | `_module` | `__init__` argument `module` | Attached `Module`, used to resolve missing inputs and dependency nodes. |
| Distribution broadcast | `_n_broadcast_samples` | `__init__` argument `n_broadcast_samples` or class default | Default Monte Carlo sample count for distribution broadcasting. |
| Vectorization | `_vectorize` | `__init__` argument `vectorize` | Requested vectorization strategy: `"auto"`, `"jax"`, or `"loop"`. |
| Execution dispatch | `_parallel` | `__init__` argument `parallel` | Loop-path parallelism setting: `False`, `True`, or explicit worker count. |
| Sampling | `_key` | `jax.random.PRNGKey(seed)` | Mutable PRNG state used for broadcast sampling. |
| Distribution broadcast | `_include_inputs` | `__init__` argument `include_inputs` | Default choice to return full `BroadcastDistribution` instead of output marginal. |
| Vectorization | `_resolved_vectorize` | Initialized to `None`, later cached by `_resolve_vectorize` | Cached result of `vectorize="auto"` probing. |
| Public facade and call resolution | `_bind` | `bind` plus extra `**kwargs` passed to `__init__` | Construction-time default inputs and configuration. |
| Public facade and call resolution | `_param_names` | Derived from `_sig.parameters` | Wrapped function parameter names excluding `self`. |
| Public facade and call resolution | `_has_var_keyword` | Derived from `_sig.parameters` | Whether the wrapped function accepts `**kwargs`. |
| Public facade and call resolution | `_child_nodes` | Inherited from `Node.__init__` | Frozen child-node mapping; usually empty for `WorkflowFunction` because no node kwargs are passed to `super().__init__()`. |
| Public facade and call resolution | `_inputs` | Inherited from `Node.__init__` | Frozen input mapping; usually empty for `WorkflowFunction` for the same reason. |

### Current Methods

| Category | Method | Current responsibility |
|---|---|---|
| Public facade and call resolution | `__init__()` | Initializes wrapper state, function metadata, binding state, broadcast defaults, execution configuration, random state, and reserved-name validation. |
| Execution dispatch | `effective_workflow_kind` | Resolves the final Prefect mode from the instance setting, global `prefect_config`, default fallback, and Prefect availability. |
| Public facade and call resolution | `_is_dependency_param(name)` | Classifies a parameter as a dependency when its annotation is a `Node` subclass. |
| Public facade and call resolution | `__call__(*args, **call_inputs)` | Main public coordinator: binds arguments, extracts reserved overrides, resolves inputs, normalizes distributions, builds the broadcast decision, dispatches execution, coerces outputs, and attaches provenance. |
| Public facade and call resolution | `_resolve_inputs(call_inputs)` | Builds final function kwargs from call-time inputs, construction-time bindings, attached module state, and function defaults; validates required and dependency-typed parameters. |
| Distribution normalization | `_convert_distributions(values)` | Converts distribution-valued arguments based on concrete `Distribution` hints or distribution protocol hints, and unwraps 0-d `DistributionArray` values when appropriate. |
| Broadcast planning | `_find_broadcast_args(values)` | Classifies values into scalar distribution broadcast arguments and array-valued sweep arguments. |
| Sampling | `_get_key()` | Splits and advances the internal JAX PRNG key. |
| Vectorization | `_resolve_vectorize(values, broadcast_args)` | Resolves `"auto"` vectorization by probing JAX traceability and caching `"jax"` or `"loop"`. |
| Broadcast planning | `_broadcast(values, dist_args, ra_args, ...)` | Dispatches among distribution-only broadcast, pure array sweep, and nested array-plus-distribution broadcast. |
| Broadcast planning | `_group_ra_args_by_parent(values, ra_args)` | Groups sweep arguments by parent identity so sibling views zip while unrelated arrays combine by product rule. |
| Sweep execution | `_slice_ra_args(values, ra_args, i, ra_groups)` | Materializes the `i`th sweep cell as a concrete function-call dictionary. |
| Sweep execution | `_execute_sweep_rows(...)` | Executes pure array sweep rows through a limited JAX `vmap` path or through loop-style execution. |
| Provenance assembly | `_make_sweep_provenance(...)` | Builds provenance metadata for pure sweep and nested broadcast outputs. |
| Distribution broadcast | `_broadcast_distributions_only(...)` | Runs scalar distribution broadcasting through empirical enumeration, JAX `vmap`, or Monte Carlo sampling; returns either a `BroadcastDistribution` or output marginal. |
| Sampling | `_sample_broadcast_args(...)` | Samples broadcast distribution arguments while preserving correlation for sibling distribution views from the same parent. |
| Distribution broadcast | `_broadcast_jax(...)` | Executes distribution broadcast through local JAX `vmap`, optionally wrapped by Prefect task or flow orchestration. |
| Distribution broadcast | `_broadcast_enumerate(...)` | Enumerates eligible `EmpiricalDistribution` sample products, propagates weights, and mixes in sampled non-empirical arguments when needed. |
| Distribution broadcast | `_broadcast_sample(...)` | Runs ordinary Monte Carlo distribution broadcast by building one call dictionary per sample. |
| Execution dispatch | `_execute_many(call_value_list)` | Executes an ordered list of call dictionaries via Prefect task mapping, Prefect flow mapping, thread pool execution, or sequential Python. |
| Execution dispatch | `_execute_many_threaded(call_value_list)` | Executes call dictionaries with `ThreadPoolExecutor`. |
| Execution dispatch | `_map_task(call_value_list, task_name=None)` | Creates a Prefect task around the wrapped function and maps it over call dictionaries. |
| Execution dispatch | `_execute_many_prefect_task(call_value_list)` | Implements `WorkflowKind.TASK` by creating a lightweight flow that hosts mapped task execution. |
| Execution dispatch | `_execute_many_prefect_flow(call_value_list)` | Implements `WorkflowKind.FLOW` by creating a named flow around mapped task execution. |

## Design Principles

1. Preserve public API compatibility.
2. Move implementation detail into private modules only.
3. Prefer deep modules over shallow forwarding.
4. Separate call resolution, planning, execution, and result assembly.
5. Execution adapters should not know probabilistic semantics.
6. Keep dependency direction from getting worse.
7. Keep scheduler-specific behavior out of this refactor.
8. Preserve JAX semantics.

### 1. Call resolution module

Add `probpipe/core/_workflow_call.py`.

Move or introduce:

- signature metadata caching;
- positional-argument binding;
- `**kwargs` unpacking;
- reserved call-time override extraction;
- input resolution from call-time inputs, construction-time binds, attached
  `Module` child nodes or inputs, and function defaults;
- required-parameter validation;
- dependency-typed parameter validation.

Proposed internal data shapes:

```python
@dataclass(frozen=True)
class WorkflowSignatureInfo:
    signature: inspect.Signature
    hints: Mapping[str, Any]
    param_names: tuple[str, ...]
    has_var_keyword: bool
    reserved_names: frozenset[str]


@dataclass(frozen=True)
class WorkflowCallOverrides:
    n_broadcast_samples: int
    include_inputs: bool
    seed: int | None


@dataclass(frozen=True)
class ResolvedWorkflowCall:
    values: dict[str, Any]
    overrides: WorkflowCallOverrides
```

Reasoning:

- Input resolution is neither broadcast planning nor execution.
- A future execution adapter should not depend on `WorkflowFunction._sig`,
  `WorkflowFunction._bind`, or `WorkflowFunction._module`.
- This module gives direct tests for function-call behavior that is currently
  hidden inside `__call__`.

### 2. Result contract module

Add `probpipe/core/_workflow_result.py`.

Move:

- `BroadcastMode`
- `BROADCAST_WRAP`
- `BROADCAST_STACK`
- `BROADCAST_NESTED`
- `_wrap_as_record`
- `_coerce_output`

Interface:

```python
def coerce_workflow_output(
    value: Any,
    *,
    broadcast_mode: BroadcastMode,
    provenance: Provenance | None,
    field_name: str,
) -> Any:
    ...
```

The exact function name may stay private (`_coerce_output`) during the first
implementation PR to minimize churn. `probpipe.core.node` can re-export the old
private helper temporarily for existing tests.

Reasoning:

- The output contract is a stable domain rule and is already mostly isolated.
- It is low-risk and provides a safe early refactor.
- Remote or distributed workers should return raw function results; result
  coercion should stay in the coordinator path after ordered results are
  collected.

### 3. Execution module

Add `probpipe/core/_workflow_execution.py`.

Move:

- `_execute_many`
- `_execute_many_threaded`
- `_map_task`
- `_execute_many_prefect_task`
- `_execute_many_prefect_flow`

Proposed internal data shapes:

```python
@dataclass(frozen=True)
class WorkflowExecutionConfig:
    mode: Literal[
        "sequential",
        "thread",
        "prefect_task",
        "prefect_flow",
    ]
    parallel: bool | int = False
    name: str = "workflow"
    prefect_task_runner: Any | None = None


@dataclass(frozen=True)
class WorkflowExecutionRequest:
    func: Callable[..., Any]
    call_value_list: list[dict[str, Any]]
    execution: WorkflowExecutionConfig
```

Recommended implementation style:

```python
class WorkflowExecutor(Protocol):
    def execute_many(self, request: WorkflowExecutionRequest) -> list[Any]:
        ...


class SequentialExecutor: ...
class ThreadExecutor: ...
class PrefectTaskExecutor: ...
class PrefectFlowExecutor: ...


def execute_many(request: WorkflowExecutionRequest) -> list[Any]:
    return resolve_executor(request.execution).execute_many(request)
```

The public internal entry point can remain function-like, but adapter classes
keep the module extensible.

Important constraints:

- The execution module must not accept or close over a `WorkflowFunction`
  instance.
- The execution module should not import JAX, `Distribution`, `Record`,
  `RecordArray`, or scheduler-specific packages such as Ray.
- The request should contain resolved execution mode, not raw
  `WorkflowKind.DEFAULT`.
- Prefect helper closures should capture only the wrapped function, call rows,
  names, and task runner metadata.

Reasoning:

- This is the highest-leverage split for backend-neutral execution.
- Sequential, thread, Prefect task, and Prefect flow dispatch are execution
  adapters over the same conceptual interface: ordered execution of many
  independent calls.
- Future execution adapters can be added behind the same interface without
  touching broadcast planning, sampling, output coercion, or provenance.
- Avoiding `self` capture improves distributed serialization behavior.

### 4. Distribution normalization module

Add `probpipe/core/_workflow_normalize.py`.

Move or introduce:

- scalar distribution conversion based on concrete `Distribution` hints;
- protocol-based distribution conversion;
- `DistributionArray` skip behavior for non-scalar arrays;
- 0-d `DistributionArray` unwrapping behavior;
- external distribution auto-conversion that currently happens inside broadcast
  argument classification.

Interface sketch:

```python
def normalize_workflow_values(
    *,
    values: dict[str, Any],
    hints: Mapping[str, Any],
) -> dict[str, Any]:
    ...
```

Reasoning:

- Current broadcast classification is partly impure because it can mutate values
  by converting external distributions to `NumericRecordDistribution`.
- Planning should be a pure classification step. Normalization should own value
  conversion.
- Centralizing `converter_registry` here preserves the existing dependency in
  one place and avoids spreading it across planning and execution modules. This
  does not attempt to resolve the broader `core -> converters` dependency
  direction.

### 5. Broadcast plan module

Add `probpipe/core/_workflow_plan.py`.

Move or introduce:

- broadcast argument classification currently handled by `_find_broadcast_args`;
- parent grouping for array-valued sweep arguments;
- sweep batch shape and total sweep size calculation;
- regime selection.

Proposed internal data shapes:

```python
@dataclass(frozen=True)
class ArrayBroadcastGroup:
    arg_names: tuple[str, ...]
    batch_shape: tuple[int, ...]
    size: int


@dataclass(frozen=True)
class BroadcastPlan:
    regime: Literal["none", "distribution", "sweep", "nested"]
    dist_args: tuple[str, ...]
    array_args: tuple[str, ...]
    array_groups: tuple[ArrayBroadcastGroup, ...]
    sweep_batch_shape: tuple[int, ...]
    n_sweep: int
    output_order: Literal["row_major"] = "row_major"
```

Reasoning:

- `WorkflowFunction` should first decide which regime applies, then execute that
  regime.
- An explicit plan makes product-rule sweep behavior easier to test directly.
- The planner should not mutate `values` and should not import the converter
  registry.

### 6. Sweep execution module

Add `probpipe/core/_workflow_sweep.py`.

Move:

- `_slice_ra_args`
- `_execute_sweep_rows`
- `_make_sweep_provenance`
- pure sweep execution from `_broadcast`
- nested outer sweep execution from `_broadcast`

Interface sketch:

```python
def execute_sweep(
    *,
    func: Callable[..., Any],
    values: dict[str, Any],
    plan: BroadcastPlan,
    execution: WorkflowExecutionConfig,
    vectorize: str,
    field_name: str,
) -> Any:
    ...
```

The exact interface should be refined during implementation. The important
constraint is that row execution uses the execution module rather than calling
the wrapped function directly, except for the existing JAX `vmap` path.

Reasoning:

- Sweep semantics are distinct from distribution marginalization semantics.
- `RecordArray` and `DistributionArray` product-shape logic should be tested as
  a coherent implementation.
- Future distributed backends may operate at this layer for parameter sweeps or
  bootstrap-like workloads, possibly with chunked row execution. Chunking itself
  is out of scope for this refactor.

## Acceptance Criteria

The full refactor is complete when:

- `WorkflowFunction` remains the public facade and preserves all existing public
  imports.
- `node.py` no longer contains the implementation details for call resolution, distribution normalization, output coercion, execution dispatch, broadcast planning, sweep execution, sampling, vectorization, and distribution-only
  broadcast.
- Existing user-visible behavior is unchanged.
- Prefect orchestration remains opt-in.
- Short-term Ray via Prefect continues to work through `prefect_config`.
- No new external dependencies are added.
- Focused tests for broadcasting, Prefect orchestration, provenance, `RecordArray` views, designs, transition helpers, validation, and output coercion pass.
- New execution tests cover result ordering, empty call lists, thread dispatch, invalid `parallel` values, and Prefect `.map()` argument expansion.
- New call-resolution tests cover positional arguments, `**kwargs`, reserved overrides, bind/module/default precedence, and dependency validation.
- New planning tests cover broadcast regime selection, sibling-view grouping, product-rule sweep shape, and non-mutating plan construction.
- Execution request objects do not contain a `WorkflowFunction` instance.
- Loop-based broadcast and sweep execution go through the execution module.
- JAX-vmap paths remain local vectorized paths and do not go through the loop execution adapter.
- The implementation leaves a clear internal execution adapter point suitable for future execution backends, including a separately designed native Ray adapter.

## Tasks

- [x] Stage 0: Characterization tests before moving behavior. (#202)
- [x] Stage 1: Result contract extraction. (#206)
- [x] Stage 2: Execution extraction. (#211)
- [x] Stage 3: Call resolution extraction. (#218)
- [x] Stage 4: Distribution normalization and BroadcastPlan introduction. (#220)
- [x] Stage 5: Sweep execution extraction. (#221)
- [x] Clean-up: Clean all wrapers. (#221)

**Comment by SoymilkWinsAgain** [comment]
Renamed the branch from `dev/workflowfunction-class-refactor` to `dev/workflowfunction-class-refactor-pr0` in order to avoid possible chaos. This unexpectedly led to the closure of PR #201. PR #202 has since been opened which is just the unfinished parts of PR#201.

**Comment by SoymilkWinsAgain** [comment]
Stage 0 finished, see #202. Stage 1 is ongoing at #206.

**Comment by SoymilkWinsAgain** [comment]
Stage 1 finished, see #206. 

Additional legacy issue: retaining compatibility aliases. These are necessary in #206 because there are some existing tests that do use something like `from probpipe.core.node import _coerce_output`. Removing these aliases would require much larger code changes, which is far beyond the scope of #206. Plan to address that in a future clean-up-dedicated PR.

**Comment by SoymilkWinsAgain** [comment]
Stage 2 is done, see #211.

Switch to #213 before workibg on next stage, because it is logically prior.

**Comment by SoymilkWinsAgain** [comment]
Stage 3 is done, see: #218.

**Comment by SoymilkWinsAgain** [comment]
Stage 4 is done. see #220. Stage 5 and the clean-up stage will be implemented in the same branch

**Comment by SoymilkWinsAgain** [comment]
Distribution broadcast extraction (not described in this issue but included in the original plan in #194) will be addressed as a legacy issue in the future because it requires a larger-scale reconfiguration, not just a copy-paste solution. It also may conflict with the upcoming #117 PR-2, and therefore should be scheduled separately.

**Comment by SoymilkWinsAgain** [comment]
Stage 5 & clean-up is done, see #221.
